timemachine/fe/rbfe.py

```diff
@@ -0,0 +1,423 @@
+
+    host_config: HostConfig
+        Configuration of the host
+
+    temperature: float
+        Temperature of the system, in Kelvins
+
+    lambda_schedule: np.array(L,)
+        Sequence of lambda windows
+
+    n_frames: int
+        Number of samples to generate per window
+
+    Returns
+    -------
+    3-tuple of all_frames, all_boxes, all_U_fns
+        all_frames has shape (L,T,N,3), all_boxes has shape (L,T,3,3), all_U_fns has shape (L,)
+
+    """
+    host_bps, host_masses = openmm_deserializer.deserialize_system(host_config.omm_system, cutoff=1.2)
+    host_conf = minimizer.minimize_host_4d(
+        [st.mol_a, st.mol_b],
+        host_config.omm_system,
+        host_config.conf,
+        st.ff,
+        host_config.box,
+    )
+
+    all_U_fns = []
+    all_frames = []
+    all_boxes = []
+
+    for lamb_idx, lamb in enumerate(lambda_schedule):
+        hgs = st.combine_with_host(convert_bps_into_system(host_bps), lamb=lamb)
+        all_U_fns.append(hgs.get_U_fns())
+        bound_impls = [p.bound_impl(np.float32) for p in hgs.get_U_fns()]
+        # minimize water box around the ligand by 4D-decoupling
+        mol_a_conf = get_romol_conf(st.mol_a)
+        mol_b_conf = get_romol_conf(st.mol_b)
+        ligand_conf = st.combine_confs(mol_a_conf, mol_b_conf)
+        combined_conf = np.concatenate([host_conf, ligand_conf])
+        x0 = combined_conf
+        v0 = np.zeros_like(x0)
+        box0 = host_config.box
+        group_idxs = get_group_indices(get_bond_list(hgs.bond))
+        run_seed = seed + lamb_idx
+        combined_masses = np.concatenate([host_masses, st.combine_masses()])
+        frames, boxes = sample(
+            bound_impls, combined_masses, temperature, x0, v0, box0, lamb, group_idxs, run_seed, n_frames=n_frames
+        )
+        all_frames.append(frames)
+        all_boxes.append(boxes)
+
+    return all_frames, all_boxes, all_U_fns
+
+
+def plot_BAR(df, df_err, fwd_delta_u, rev_delta_u, title, plot_idx):
+    """
+    Generate a subplot showing overlap for a particular pair of delta_us.
+
+    Parameters
+    ----------
+    df: float
+        reduced free energy
+
+    df_err: float
+        reduced free energy error
+
+    fwd_delta_u: array
+        reduced works
+
+    rev_delta_u: array
+        reduced reverse works
+
+    title: str
+        title to use
+
+    plot_idx: triple (n_row, n_col, n_pos)
+        where to place the subplot
+
+    """
+    plt.subplot(*plot_idx)
+    plt.title(f"{title}, dg: {df:.2f} +- {df_err:.2f} kTs")
+    plt.hist(fwd_delta_u, alpha=0.5, label="fwd", density=True, bins=20)
+    plt.hist(-rev_delta_u, alpha=0.5, label="-rev", density=True, bins=20)
+    plt.xlabel("work (kTs)")
+    plt.legend()
+
+
+SimulationResult = namedtuple(
+    "SimulationResult", ["all_dGs", "all_errs", "plot_png", "frames", "boxes", "U_fns", "lambda_schedule"]
+)
+
+
+def estimate_free_energy_given_samples(all_frames, all_boxes, all_U_fns, temperature, lambda_schedule, prefix):
```

Comment:
1. Is this implementation RBFE-specific, or can it be moved somewhere less specific, like `estimator.py` or similar?

(Implementation seems equally applicable to other cases where a free energy can be estimated from equilibrium samples.)

2. Multiple estimators are possible in this setting -- may want to mention in name or docstring which estimator this function implements.

Reply:
1) the problem with `estimator.py` is that it's fairly tied to RABFE specific logic (like setting up the orientational restraints), didn't seem right to put it there
2) will add docstring

Reply:
will deal with 1) during deboggling